clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy

```diff
@@ -243,6 +247,124 @@ class AzureNetworkClient extends AzureBaseClient {
+    try {
+      def currentTime = System.currentTimeMillis()
+      def appGateway = executeOp({appGatewayOps.get(resourceGroupName, appGatewayName)})?.body
+      if (appGateway) {
+        def agItem = AzureAppGatewayDescription.getDescriptionForAppGateway(appGateway)
+        agItem.appName = AzureUtilities.getAppNameFromResourceId(appGateway.id)
+        agItem.tags = appGateway.tags
+        agItem.dnsName = getDnsNameForAppGateway(AzureUtilities.getResourceGroupNameFromResourceId(appGateway.id), appGateway.name)
+        agItem.lastReadTime = currentTime
+        return agItem
+      }
+    } catch (CloudException e) {
+      log.error("getAppGateway(${resourceGroupName},${appGatewayName}) -> Cloud Exception ", e)
+    } catch (Exception e) {
+      log.error("getAppGateway(${resourceGroupName},${appGatewayName}) -> Exception ", e)
+    }
+
+    null
+  }
+
+  /**
+   * Retrieve a collection of all application gateways for a give set of credentials and the location
+   * @param region the location of the virtual network
+   * @return a Collection of objects which represent an Application Gateway in Azure
+   */
+  Collection<AzureAppGatewayDescription> getAppGatewaysAll(String region) {
+    def result = new ArrayList<AzureAppGatewayDescription>()
+
+    try {
+      def currentTime = System.currentTimeMillis()
+      def appGateways = executeOp({appGatewayOps.listAll()})?.body
+
+      appGateways.each {item ->
+        if (item.location == region) {
+          try {
+            def agItem = AzureAppGatewayDescription.getDescriptionForAppGateway(item)
+            agItem.appName = AzureUtilities.getAppNameFromResourceId(item.id)
+            agItem.tags = item.tags
+            agItem.dnsName = getDnsNameForAppGateway(AzureUtilities.getResourceGroupNameFromResourceId(item.id), item.name)
+            agItem.lastReadTime = currentTime
+            result += agItem
+          } catch (RuntimeException re) {
+            // if we get a runtime exception here, log it but keep processing the rest of the
+            // load balancers
+            log.error("Unable to process load balancer ${item.name}: ${re.message}")
+          }
+        }
+      }
+    } catch (Exception e) {
+      log.error("getAppGatewaysAll -> Unexpected exception ", e)
+    }
+
+    result
+  }
+
+  /**
+   * get the dns name associated with an Application Gateway in Azure
+   * @param resourceGroupName name of the resource group where the application gateway was created (see application name and region/location)
+   * @param appGatewayName the name of the application gateway in Azure
+   * @return the dns name of the Public IP for the given application gateway or "dns-not-found"
+   */
+  String getDnsNameForAppGateway(String resourceGroupName, String appGatewayName) {
+    String dnsName = "dns-not-found"
+
+    try {
+      def appGateway = executeOp({appGatewayOps.get(resourceGroupName, appGatewayName)})?.body
```

Comment:
In both instances where you call this method, you've just retrieved the App Gateway from Azure via the list operation. Why do you need to retrieve again?